src/inferenceCUDA.jl

```diff
@@ -0,0 +1,349 @@
+        end
+
+        # Step 2: Normalize and scale
+        ss /= length_x
+        ss += 1f-6
+        ss = 1f0 / sqrt(ss)
+        o[i] = weight[i] * (ss * x[i])
+    end
+    nothing
+end
+function rmsnorm!(o::CuVector{Float32}, x::CuVector{Float32}, weight::CuVector{Float32})
+  length_x = length(x)
+
+  threads_per_block = 256
+  blocks_per_grid = ceil(Int, length_x / threads_per_block)
+
+  @cuda threads=threads_per_block blocks=blocks_per_grid rmsnorm_kernel(o, x, weight, length_x)
+end
+
+
+@views function transformer!(token::Int, pos::Int, config::ModelConfig, s::RunStateCUDA, weights::TransformerWeightsCUDA)
+  x = s.x
+
+  (;
+      dim,
+      hidden_dim,
+      n_layers,
+      n_heads,
+      n_kv_heads,
+  ) = config
+
+  head_size = dim ÷ n_heads
+  CUDA.@sync begin # syncing to make sure timings are accurate outside. Although this is not important.
+  # copy the token embedding into x
+  dequantize_cuda!(x, weights.token_embedding_table[:, token])
+
+  # forward all the layers
+  for l in 1:n_layers
+      w = weights.layers[l]
+      kv = s.kvcache_layers[l]
+
+      # attention rmsnorm
+      @time CUDA.@sync rmsnorm!(s.xb, x, w.rms_att_weight)
+
+      # qkv matmuls for this position
+      @time CUDA.@sync matmul!(s.q, w.wq, s.xb)
+      matmul!(s.k, w.wk, s.xb)
+      matmul!(s.v, w.wv, s.xb)
+      
+      q = reshape(s.q, head_size, n_heads)
+      k = reshape(s.k, head_size, n_kv_heads)
+      
+      # apply RoPE rotation to the q and k vectors for each head
+      @time CUDA.@sync rope!(q, pos, config)
+      rope!(k, pos, config)
+
+      # save key,value at this time step (pos) to our kv cache
+      @time CUDA.@sync copyto!(kv.key_cache[:, :, pos], s.k)
+      copyto!(kv.value_cache[pos, :, :], s.v)
+
+      # take a contiguous slice of the attention buffer
+      att = reshape(s.att[1:(n_heads*pos)], pos, n_heads)
+
+      # multihead attention
+      @time CUDA.@sync attention_weights!(att, kv.key_cache, q)
+
+      att ./= sqrt(Float32(head_size))
+
+      softmax_for!(att, n_heads)
+
+      xb = reshape(s.xb, head_size, n_heads)
+
+      # weighted sum of the values
+      combine_values!(xb, kv.value_cache, att)
+
+      # final matmul to get the output of the attention
+      matmul!(s.xb2, w.wo, s.xb)
+
+      # residual connection back into x
+      x .+= s.xb2
+
+      # ffn rmsnorm
+      rmsnorm!(s.xb, x, w.rms_ffn_weight)
+
+      # Now for FFN in PyTorch we have: self.w2(F.silu(self.w1(x)) * self.w3(x))
+      # first calculate self.w1(x) and self.w3(x)
+      matmul!(s.hb, w.w1, s.xb)
+      matmul!(s.hb2, w.w3, s.xb)
+
+      s.hb .= silu.(s.hb)
+      
+      s.hb .*= s.hb2
+
+      # final matmul to get the output of the ffn
+      matmul!(s.xb, w.w2, s.hb)
+      # residual connection
+      x .+= s.xb
+  end
+
+  # final rmsnorm
+  rmsnorm!(x, x, weights.rms_final_weight)
+
+  # classifier into logits
+  matmul!(s.logits, weights.output_weight, x)
+  
+  end
+  return nothing
+end
```

Comment:
```suggestion
end

```

Reply:
(adding a newline at the end of the file)

Reply:
Okay, going to keep this pattern. :)